harness/determined/exec/tensorboard.py

```diff
@@ -161,17 +181,85 @@ def start_tensorboard(
 
 
 class TBFetchManager:
+    """Simple Container Class to manage the state of the TensorBoard remote fetchers"""
+
     def __init__(self) -> None:
+        self._lock = threading.Lock()
         self._ready = False
-        self.num_fetched_files = 0
+        self._num_fetched_files = 0
 
-    # TODO: If we support multi-threaded fetching in the future, this will
-    # need a lock
     def on_file_fetched(self) -> None:
-        if not self._ready:
-            self._ready = True
-            print(TENSORBOARD_TRIGGER_READY_MSG, flush=True)
-        self.num_fetched_files += 1
+        """Runs on each callback from a fetcher to update the state.
+        We delay signalling the READY_SIGNAL to give TensorBoard a moment
+        to do its internal book keeping and prevent the "no data"
+        issues we were previously seeing when signalling ready immediately
+        after the first file is fetched.
+
+        TensorBoard >=v2.5.0 has a "load_fast=true" flag to alleviate this problem
+        but for backwards compatibility reasons, we cannot use it yet.
+        More info here: https://github.com/tensorflow/tensorboard/issues/4784
+        """
+        with self._lock:
+            if not self._ready:
+                self._ready = True
```

Comment:
what is the purpose of this `_ready` flag?

Reply:
It's to make sure that we send the ready signal exactly once after the first file is synced

Reply:
oh nvm, i completely misread the indent blocks here. gotcha